File: src/com/douma/_13_day_综合应用一/_170/_170_TwoSum2.cpp
/**
* 抖码算法，让算法学习变的简单有趣
* @作者 : 老汤
*/

#include <unordered_map>

using namespace std;

class _170_TwoSum2 {
private:
    unordered_map<int, int> nums;

public:

    _170_TwoSum2() {}

    void add(int number) {
        if (nums.find(number) != nums.end()) {
            nums[number] += 1;
        } else {
            nums[number] = 1;
        }
    }

    bool find(int value) {
        for (auto& kv : nums) {
            int target = value - kv.first;
            if (target == kv.first && nums[target] > 1) return true;
            if (target != kv.first && nums.count(target)) return true;
        }
        return false;
    }
};

File: src/com/douma/_13_day_综合应用一/_170/_170_TwoSum2.js
// 抖码算法，让算法学习变的简单有趣
// 作者：老汤

class TwoSum {
    constructor() {
        this.nums = new Map()
    }

    add(number) {
        if (this.nums.has(number)) {
            this.nums.set(number, this.nums.get(number) + 1)
        } else {
            this.nums.set(number, 1)
        }
    }

    find(value) {
        for (const key of this.nums.keys()) {
            const target = value - key
            if (target == key && this.nums.get(target) > 1) return true
            if (target != key && this.nums.has(target)) return true
        }
        return false
    }
}